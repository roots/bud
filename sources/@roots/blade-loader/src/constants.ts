export const scriptLoaders = [
  {loader: `@roots/blade-loader/script-loader`, type: `asset/source`},
  {loader: `@roots/blade-loader/asset-loader`, type: `asset/source`},
]

export const scriptPattern = /\.(js|mjs|jsx|ts|tsx|php)$/
