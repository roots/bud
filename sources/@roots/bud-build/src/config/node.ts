import type {Factory} from './index.js'

export const node: Factory<`node`> = async ({hooks}) =>
  hooks.filter(`build.node`, false)
