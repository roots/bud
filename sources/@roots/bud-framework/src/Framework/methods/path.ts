import {join} from 'path'

import {Framework, Locations} from '../..'

export interface path {
  (
    this: Framework,
    key: keyof Locations & string,
    ...path: string[]
  ): string
}

export interface path {
  (key: `${keyof Locations & string}`, ...path: string[]): string
}

export const path: path = function (key, ...path): string {
  const project = this.hooks.filter(`location.project`)
  const partial = this.hooks.filter(`location.${key}`)

  const useAbsolute = project === partial || partial.startsWith('/')

  return useAbsolute
    ? join(...[partial, ...(path ?? [])].filter(Boolean))
    : join(...[project, partial, ...(path ?? [])].filter(Boolean))
}
