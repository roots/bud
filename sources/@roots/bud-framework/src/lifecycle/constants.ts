/**
 * Service lifecycle events
 *
 * @public
 */
export const LIFECYCLE_EVENTS = [
  `bootstrap`,
  `bootstrapped`,
  `register`,
  `registered`,
  `boot`,
  `booted`,
]

/**
 * Services which are only instantiated in the parent compiler context.
 *
 * @public
 */
export const PARENT_SERVICES: Array<string> = [
  `@roots/bud-compiler`,
  `@roots/bud-dashboard`,
  `@roots/bud-server`,
]

/**
 * Services which are only instantiated in development
 *
 * @public
 */
export const DEVELOPMENT_SERVICES: Array<string> = [`@roots/bud-server`]
