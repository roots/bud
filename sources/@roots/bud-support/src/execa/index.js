import {execa} from 'execa'
export default execa
