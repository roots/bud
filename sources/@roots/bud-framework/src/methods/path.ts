import type {Bud} from '@roots/bud-framework'

import {isAbsolute, join, resolve, sep} from 'node:path'

/**
 * Transform `@alias` path
 *
 * @param app - Bud instance
 * @param base - Path segment
 * @returns string
 */
export interface path {
  (...values: Array<string>): string
}
export const path: path = function (...values) {
  /* Exit early with context.basedir if no path was passed */
  if (!values?.length) return this.context.basedir

  if (isAbsolute(join(...values))) {
    return join(...values)
  }

  values = values.flatMap(value => value.split(sep))

  const hash =
    this.hooks.filter(`value.hashFormat`, `[contenthash:6]`) ??
    `[contenthash:6]`
  const name = this.hooks.filter(`value.fileFormat`, `[name]`) ?? `[name]`

  const transformMagicString = makeParseMagicString(
    this.context.hash ? `${name}.${hash}` : name,
    hash,
  )

  const parseAlias = makeParseAlias(this)

  const result = join(...values.map(transformMagicString).map(parseAlias))

  return isAbsolute(result)
    ? result
    : resolve(this.context.basedir, result)
}

/**
 * Transform `@alias` path
 *
 * @param app - Bud instance
 * @param base - Path segment(s)
 * @returns string
 */
export const makeParseAlias =
  (app: Bud) =>
  (segment: string, i: number): string => {
    if (!(`location.${segment}` in app.hooks.syncStore.store))
      return segment

    return app.hooks.filter(`location.${segment}` as any)
  }

export const makeParseMagicString =
  (name: string, hash: string) => (segment: string) => {
    return segment
      .replace(`@file`, `@path@base`)
      .replace(`@base`, `@name@ext`)
      .replace(`@name`, name)
      .replace(`@hash`, hash)
      .replace(`@path`, `[path]`)
      .replace(`@ext`, `[ext]`)
  }
