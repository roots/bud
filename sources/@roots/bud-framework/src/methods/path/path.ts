import {isAbsolute, join, resolve, sep} from 'node:path'

import type {Bud} from '../../bud.js'

/**
 * Transform `@alias` path
 *
 * @param app - Bud instance
 * @param base - Path segment
 * @returns string
 */
export interface path {
  (...values: Array<string>): string
}
export const path: path = function (...values) {
  const app = this as Bud

  /* Exit early with context.basedir if no path was passed */
  if (!values?.length) return app.context.basedir

  if (isAbsolute(join(...values))) {
    return join(...values)
  }

  values = values.flatMap(value => value.split(sep))

  const hash = app.hooks.filter(`value.hashFormat`, `[contenthash:6]`)
  const name = app.hooks.filter(`value.fileFormat`, `[name]`)

  const transformMagicString = makeParseMagicString(
    app.hooks.filter(`feature.hash`) ? `${name}.${hash}` : name,
    hash,
  )

  const parseAlias = makeParseAlias(app)

  const result = join(...values.map(transformMagicString).map(parseAlias))

  return isAbsolute(result) ? result : resolve(app.context.basedir, result)
}

/**
 * Transform `@alias` path
 *
 * @param app - Bud instance
 * @param base - Path segment(s)
 * @returns string
 */
export const makeParseAlias =
  (app: Bud) =>
  (segment: string, i: number): string => {
    if (!(`location.${segment}` in app.hooks.syncStore.store))
      return segment

    return app.hooks.filter(`location.${segment}` as any)
  }

export const makeParseMagicString =
  (name: string, hash: string) => (segment: string) => {
    return segment
      .replace(`@file`, `@path@base`)
      .replace(`@base`, `@name@ext`)
      .replace(`@name`, name)
      .replace(`@hash`, hash)
      .replace(`@path`, `[path]`)
      .replace(`@ext`, `[ext]`)
  }
