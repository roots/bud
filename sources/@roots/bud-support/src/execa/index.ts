import {execa} from 'execa'

export default execa
export * from 'execa'
