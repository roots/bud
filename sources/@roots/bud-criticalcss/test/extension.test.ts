/* eslint-disable no-console */
import {join} from 'node:path'

import {paths} from '@repo/constants'
import * as fs from 'fs-jetpack'
import {execa} from 'execa'
import {describe, expect, it} from 'vitest'

describe(`@roots/bud-criticalcss`, () => {
  it(`extract`, async () => {
    await runFixture(`extract`)
    expect(await readCritical(`extract`)).toMatchSnapshot()
    expect(await readOriginal(`extract`)).toMatchSnapshot()
  }, 60000)

  it(`no-extract`, async () => {
    await runFixture(`no-extract`)
    expect(await readCritical(`no-extract`)).toMatchSnapshot()
    expect(await readOriginal(`no-extract`)).toMatchSnapshot()
  }, 60000)
}, 60000)

const baseParts = [
  paths.sources,
  `@roots`,
  `bud-criticalcss`,
  `test`,
  `__fixtures__`,
]

const runFixture = async target =>
  await execa(
    `yarn`,
    [
      `workspace`,
      `@tests/bud-criticalcss__${target}`,
      `run`,
      `bud`,
      `build`,
      `--no-cache`,
    ],
    {reject: false, timeout: 30000},
  )

const readOriginal = async target =>
  await fs.readAsync(
    join(...baseParts, target, `dist`, `css`, `index.css`),
    `utf8`,
  )

const readCritical = async target =>
  await fs.readAsync(
    join(...baseParts, target, `dist`, `critical`, `css`, `index.css`),
    `utf8`,
  )
