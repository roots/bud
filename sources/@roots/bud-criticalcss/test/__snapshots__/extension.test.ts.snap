// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@roots/bud-criticalcss > extract 1`] = `"body{background-color:#00f}.foo{background-color:#000}"`;

exports[`@roots/bud-criticalcss > extract 2`] = `".ignore-style .test{background-color:red}"`;

exports[`@roots/bud-criticalcss > no-extract 1`] = `"body{background-color:#00f}"`;

exports[`@roots/bud-criticalcss > no-extract 2`] = `"body{background-color:blue}.foo{background-color:#000}"`;
